### PR TITLE
feature/sup-8-email

### PR DESCRIPTION
**Pull request by alesiahlushakova**

### CODE REVIEW

File: src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -45,16 +47,20 @@ public UserDto save(UserDto userDTO) {
                     userMapper.update(userDTO, existing);
                     return existing;
                 })
-                .orElseGet(() -> userMapper.map(userDTO));
+                .orElseGet(() -> {
+                    MailService mailService = new MailServiceImpl();
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java]: why its not a bean?

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            Sender sender = new Sender(USERNAME, PASSWORD);
+            Email email = new Email();
+            email.setTo(user.getEmail());
+            email.setSubject(ACTIVATION_LINK);
+            email.setText(buildMessage(user, String.format("localhost:8080/users/%s/true", user.getUsername())));
+            sender.send(email);
+    }
+
+    private String buildMessage(UserDto user, String url) {
+        ST st = new ST("Hello, <username>! To activate Your profile, please follow the link below <link>");
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: Please dont use external libs for things that already exist in Spring, use TemplateEngine, which renders HTML with Thymeleaf

File: src/main/java/by/itech/lab/supplier/service/mail/Sender.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package by.itech.lab.supplier.service.mail;
+
+import javax.mail.*;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/Sender.java]: configure wildcard imports

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    private static final String USERNAME = "adminMail";
+    private static final String PASSWORD = "password";
+    private static final String ACTIVATION_LINK = "Activation link";
+
+    public void sendMail(UserDto user) {
+            Sender sender = new Sender(USERNAME, PASSWORD);
+            Email email = new Email();
+            email.setTo(user.getEmail());
+            email.setSubject(ACTIVATION_LINK);
+            email.setText(buildMessage(user, String.format("localhost:8080/users/%s/true", user.getUsername())));
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: please create a property which will contain a path to the application

**Review by svareyko — CHANGES_REQUESTED** [review]

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java (hunk 2 above)
Comment (alesiahlushakova) [reply src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: I don't understand how to work with temlate engine here

File: src/main/java/by/itech/lab/supplier/service/mail/Email.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package by.itech.lab.supplier.service.mail;
+
+public class Email {
+    private String from;
+    private String to;
+    private String subject;
+    private String text;
+
+    public String getFrom() {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/Email.java]: what about lombok?

File: src/main/java/by/itech/lab/supplier/service/mail/MailService.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package by.itech.lab.supplier.service.mail;
+
+import by.itech.lab.supplier.domain.User;
+import by.itech.lab.supplier.dto.UserDto;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailService.java]: in the interface? really? 😃

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java (hunk 2 above)
Comment (svareyko) [reply src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]:
```
Map<String, Object> attributes = new HashMap();
attributes.put("firstName", "Alesya");
attributes.put("lastName", "Glushakova");
attributes.put("title", "Registration email");
Context context = new Context(*locale*, attributes)
templateEngine.process("path_to_html_template_in_resources(excluding_.html)", context);
```
HTML Template:
```
<!DOCTYPE html>
<html xmlns:th="http://www.thymeleaf.org">
  <head>
    <title th:text="${title}"></title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  </head>
  <body>
    <p th:text="|Hello ${firstName} ${lastName},|"></p>
  </body>
</html>
```

File: src/main/java/by/itech/lab/supplier/service/mail/Email.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package by.itech.lab.supplier.service.mail;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/Email.java:8]: is this constructor used anywhere?

File: src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -45,16 +49,19 @@ public UserDto save(UserDto userDTO) {
                     userMapper.update(userDTO, existing);
                     return existing;
                 })
-                .orElseGet(() -> userMapper.map(userDTO));
+                .orElseGet(() -> {
+                    mailService.sendMail(userDTO);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java]: what if save failed?

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        final TemplateEngine templateEngine = new TemplateEngine();
+        templateEngine.addTemplateResolver(htmlTemplateResolver());
+        return templateEngine;
+    }
+
+    private ITemplateResolver htmlTemplateResolver() {
+        final ClassLoaderTemplateResolver templateResolver = new ClassLoaderTemplateResolver();
+        templateResolver.setOrder(1);
+        templateResolver.setResolvablePatterns(Collections.singleton("html/*"));
+        templateResolver.setSuffix(".html");
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: which is default value?

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        templateResolver.setTemplateMode(TemplateMode.HTML);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: which is default value?

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+@PropertySource("classpath:application.properties")
+public class MailServiceImpl implements MailService {
+    private static final String ACTIVATION_LINK = "Activation link";
+    private static final String HTML_TEMPLATE_NAME = "html/email-simple";
+    private final TemplateEngine templateEngine;
+    private final Sender sender;
+
+    public void sendMail(final UserDto userDto) {
+        Map<String, Object> emailMap = new HashMap<>();
+        emailMap.put("firstName", "Alesya");
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: change your mockup data

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        emailMap.put("lastName", "Glushakova");
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: change your mockup data

File: src/main/resources/templates.html/email-simple.html [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+<html xmlns:th="http://www.thymeleaf.org">
+<head>
+    <title th:text="${title}"></title>
+    <meta http-equiv="Content-Type" content="text/html" charset="UTF-8">
+</head>
+<body>
+<p th:text="${firstName}"></p>
+<p th:text="${lastName}"></p>
+</body>
+</html>
```
Comment (svareyko) [inline src/main/resources/templates.html/email-simple.html]: empty line?

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import org.thymeleaf.templateresolver.ITemplateResolver;
+
+import java.util.Collections;
+import java.util.Properties;
+
+@Configuration
+@PropertySource("classpath:application.properties")
+public class MailConfig {
+    private static final String EMAIL_TEMPLATE_ENCODING = "UTF-8";
+    @Value("${admin.email}")
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: please use default value

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    private String email;
+    @Value("${admin.password}")
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: please use default value

**Review by svareyko — CHANGES_REQUESTED** [review]

File: src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java [hunk 16]
```diff
@@ -23,6 +26,8 @@
 
     private final UserMapper userMapper;
 
+    private final MailServiceImpl mailService;
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/impl/UserServiceImpl.java]: Why do you link impl but not interface?

File: src/main/java/by/itech/lab/supplier/service/UserService.java [hunk 17]
```diff
@@ -18,7 +18,7 @@
 
     void delete(Long id);
 
-    boolean changeActiveStatus(Long id, boolean status);
+    boolean changeActiveStatus(String username, boolean status);
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/UserService.java]: whats wrong with id?

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+import org.thymeleaf.TemplateEngine;
+import org.thymeleaf.templatemode.TemplateMode;
+import org.thymeleaf.templateresolver.ClassLoaderTemplateResolver;
+import org.thymeleaf.templateresolver.ITemplateResolver;
+
+import java.util.Collections;
+import java.util.Properties;
+
+@Configuration
+@PropertySource("classpath:application.properties")
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: is this required?

File: src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    public TemplateEngine emailTemplateEngine() {
+        final TemplateEngine templateEngine = new TemplateEngine();
+        templateEngine.addTemplateResolver(htmlTemplateResolver());
+        return templateEngine;
+    }
+
+    private ITemplateResolver htmlTemplateResolver() {
+        final ClassLoaderTemplateResolver templateResolver = new ClassLoaderTemplateResolver();
+        templateResolver.setOrder(1);
+        templateResolver.setResolvablePatterns(Collections.singleton("html/*"));
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailConfig.java]: is this doesnt match to the default value?

File: src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+import org.thymeleaf.TemplateEngine;
+import org.thymeleaf.context.Context;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+@AllArgsConstructor
+@Service
+@PropertySource("classpath:application.properties")
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/MailServiceImpl.java]: is this required?

File: src/main/java/by/itech/lab/supplier/service/mail/Sender.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+import java.util.Properties;
+
+public class Sender {
+    private String username;
+    private String password;
+    private Properties props;
+
+    public Sender(String username, String password, Properties props) {
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/service/mail/Sender.java]: replace with AllArgsConstructor

**Review by svareyko — CHANGES_REQUESTED** [review]

File: src/main/java/by/itech/lab/supplier/dto/UserDto.java [hunk 22]
```diff
@@ -52,4 +51,4 @@
     private Set<WayBill> driverWayBills = new HashSet<>();
     private Set<Application> creatorApplications = new HashSet<>();
     private Set<Application> updatorApplications = new HashSet<>();
-}
+}
```
Comment (svareyko) [inline src/main/java/by/itech/lab/supplier/dto/UserDto.java:54]: empty line?

**Review by svareyko — APPROVED** [review]

**Review by IrynaZazyba — APPROVED** [review]

**Pull request merged**